### PR TITLE
feat(asr): add Artifact Subspace Reconstruction module (standard, rASR, adaptive, Juggler)

### PR DESCRIPTION
**Pull request by snesmaeili**
## Summary

Adds an Artifact Subspace Reconstruction (ASR) module to `mne-denoise`. ASR
repairs short, high-amplitude, spatially-structured EEG artifacts by calibrating
a clean-data covariance model and reconstructing burst-contaminated subspaces
(clean_rawdata lookahead + moving-covariance + raised-cosine blending).

The estimators follow the package conventions: scikit-learn
`BaseEstimator`/`TransformerMixin` with `fit`/`transform`/`fit_transform`
(+ `partial_fit` for streaming), accepting MNE `Raw`/`Epochs` or NumPy arrays,
and exposing fitted `*_` attributes.

## What's included

**Variants**
- `ASR(method="standard")` — Euclidean clean_rawdata ASR (the default).
- `ASR(method="riemannian")` — MATLAB `rASRMatlab`-faithful backend (kept behind
  `experimental=True`).
- `ASR(method="riemannian_windowed")` — Riemannian robust calibration with a
  per-window eigendecomposition, so the `cutoff` knob behaves like standard ASR.
- `AdaptiveASR(variant="psp"|"psw"|"mw")` — Hebbian/anti-Hebbian adaptive ASR,
  with a moving-window mode (`mw_mode="final_state"|"sliding"`).
- `JugglerASR(strategy="dbscan"|"gev")` — pointwise-amplitude reference
  selection for high-motion / MoBI EEG.

**Unified API** across all three estimators: `get_diagnostics()`,
`get_calibration_mask()` (+ `calibration_mask_kind_`), and
`to_annotations(kind="repair"|"rejection"|"calibration")`.

**Visualization** — `mne_denoise.viz` gains 10 `plot_asr_*` helpers (overlay,
cutoff sweep, PSD comparison, variance topomap, repair timeline, calibration
fraction, component reconstruction, blink reduction, grand average, method
comparison), following the existing viz contract (`(fig, ax)`, MNE or ndarray
inputs, `ax=`/`show=`/`fname=`).

**Docs & examples** — `docs/asr.rst` narrative + API reference, and 5 runnable
gallery examples under `examples/asr/`.

## Validation

Each variant is validated against its source paper and reference implementation
(Mullen 2013, Chang 2020, Blum 2019, Tsai et al., Kim 2025), with MATLAB
cross-check fixtures for the standard, rASR, adaptive, and riemannian_windowed
backends. Reproducibility tooling lives under `scripts/`.

## Testing

- Full suite green: **699 passed, 12 skipped** (skips are MATLAB-Engine-gated
  parity tests).
- **90%** statement coverage on `mne_denoise/asr` + `mne_denoise/viz/asr.py`.
- `ruff check` and `ruff format --check` clean across the repo.

## Notes

- MATLAB parity `.mat` fixtures are generated locally and gitignored; the parity
  tests skip when the fixtures (or MATLAB) are unavailable.
- The `riemannian` backend is cutoff-invariant on real EEG by construction and
  stays behind `experimental=True`; `riemannian_windowed` is the recommended
  Riemannian path.

### CODE REVIEW

File: docs/asr/ASR_IMPLEMENTATION_PLAN.md
Comment (BabaSanfour) [inline docs/asr/ASR_IMPLEMENTATION_PLAN.md:1]: This should be removed

File: docs/changes/devel/27.feature.rst
Comment (BabaSanfour) [inline docs/changes/devel/27.feature.rst:1]: all the added files in devel should be in one file .. and it should match the PR number (#36

File: docs/changes/devel/28.feature.rst
Comment (BabaSanfour) [inline docs/changes/devel/28.feature.rst:1]: same (#36 )

File: docs/changes/devel/29.bugfix.rst
Comment (BabaSanfour) [inline docs/changes/devel/29.bugfix.rst:1]: same (#36 )

File: docs/changes/devel/30.feature.rst
Comment (BabaSanfour) [inline docs/changes/devel/30.feature.rst:1]: same (#36

File: docs/changes/devel/31.feature.rst
Comment (BabaSanfour) [inline docs/changes/devel/31.feature.rst:1]: same (#36

File: examples/asr/plot_01_asr_basics.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+print(f"Calibration windows kept: {asr.clean_window_mask_.sum()}")
+print(f"Repaired windows: {(asr.n_components_reconstructed_ > 0).sum()}")
+print(f"Repaired sample fraction: {asr.fraction_reconstructed_samples_:.2%}")
+metrics = compute_asr_qa_metrics(data, clean, asr)
+print(f"Variance removed: {metrics['variance_removed_pct']:.2f}%")
+
+# %%
+# Plot One Channel
+# ----------------
+fig, ax = plt.subplots(figsize=(10, 4))
```
Comment (BabaSanfour) [inline examples/asr/plot_01_asr_basics.py]: why not the viz functions we have?

File: examples/asr/plot_02_mne_raw_qc.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+print(f"Variance removed: {metrics['variance_removed_pct']:.2f}%")
+
+# %%
+# Plot One EEG Channel
+# --------------------
+channel = "EEG 00"
+noisy = raw.get_data(picks=[channel])[0]
+clean = raw_clean.get_data(picks=[channel])[0]
+
+fig, ax = plt.subplots(figsize=(11, 4))
```
Comment (BabaSanfour) [inline examples/asr/plot_02_mne_raw_qc.py]: same here, why not use the viz module you created ? probably in all examples.

File: examples/tutorials/paper4_asr_kothe_chang.ipynb
Comment (BabaSanfour) [inline examples/tutorials/paper4_asr_kothe_chang.ipynb:1]: This should be moved to the tutorials branch and not here.

File: mne_denoise/asr/adaptive.py [hunk 3]
```diff
@@ -0,0 +1,1580 @@
+"""Adaptive ASR variants backed by the AASR MATLAB reference.
```
Comment (BabaSanfour) [inline mne_denoise/asr/adaptive.py]: Needs to be improved

File: mne_denoise/asr/adaptive.py [hunk 4]
```diff
@@ -0,0 +1,1580 @@
...
+    _regularize_spd,
+    _resolve_max_dims_clean_rawdata,
+    _round_half_up,
+    _sample_mask_from_removed_windows,
+    _sqrtm_spd,
+    _validate_array_2d,
+    fit_eeg_distribution,
+)
+
+try:
```
Comment (BabaSanfour) [inline mne_denoise/asr/adaptive.py:57]: Is this how we handle mne imports in the other modules? the API should be consistent

File: mne_denoise/asr/adaptive.py [hunk 5]
```diff
@@ -0,0 +1,1580 @@
...
+    import mne
+    from mne.epochs import BaseEpochs
+    from mne.evoked import Evoked
+    from mne.io import BaseRaw
+except ImportError:  # pragma: no cover - MNE is a required project dependency
+    mne = None
+
+
+_AASR_BETA_GRID = np.arange(1.7, 3.5 + 1e-12, 0.15, dtype=np.float64)
```
Comment (BabaSanfour) [inline mne_denoise/asr/adaptive.py]: I haven't checked these yet (will do after your changes), but make sure each function is needed (if it can be simplified / inlined that would be helpful) .. also its making the script long.. maybe some of them could live in a utils script

File: mne_denoise/asr/adaptive.py [hunk 6]
```diff
@@ -0,0 +1,1580 @@
...
+        "cov": cov_state.copy() if cov_state is not None else None,
+        "carry": carry_out,
+        "iir": iir_state.copy() if iir_state is not None else None,
+        "last_R": last_R.copy(),
+        "last_trivial": bool(last_trivial),
+    }
+    return X_clean, diagnostics, next_state
+
+
+class AdaptiveASR(ASR):
```
Comment (BabaSanfour) [inline mne_denoise/asr/adaptive.py:68]: from a first glance looks fine, but did you make sure to optimize the code ? to avoid memory / speed issues?

File: mne_denoise/asr/core.py [hunk 7]
```diff
@@ -0,0 +1,3059 @@
...
+        "cov_condition": float(np.linalg.cond(C)),
+        "covariance_geometry": method,
+        "filter_kind": filter_kind,
+    }
+    diagnostics.update(memory_info)
+    diagnostics.update(riemannian_info)
+    return state, diagnostics
+
+
+def fit_eeg_distribution(
```
Comment (BabaSanfour) [inline mne_denoise/asr/core.py]: does it work on meg ?

File: mne_denoise/asr/core.py [hunk 8]
```diff
@@ -0,0 +1,3059 @@
...
+                chunk_samples=n_stream_input,
+                used_memory_bound=False,
+            )
+        )
+    if store_reconstruction_matrices:
+        diagnostics["reconstruction_matrices"] = np.asarray(reconstruction_matrices)
+    return X_clean, diagnostics
+
+
+class ASR(BaseEstimator, TransformerMixin):
```
Comment (BabaSanfour) [inline mne_denoise/asr/core.py]: same comment as adaptive ASR:
1. helper functions; are they needed could they be inlined/ simplified? 
2. was this optimized for performance in terms of memory / speed?

File: mne_denoise/asr/core.py [hunk 9]
```diff
@@ -0,0 +1,3059 @@
...
+    for start, stop in spans[1:]:
+        last_start, last_stop = merged[-1]
+        if start <= last_stop:
+            merged[-1] = (last_start, max(last_stop, stop))
+        else:
+            merged.append((start, stop))
+    return merged
+
+
+def _mask_to_sample_spans(mask: np.ndarray) -> list[tuple[int, int]]:
```
Comment (BabaSanfour) [inline mne_denoise/asr/core.py]: too long script, can benefit from being split into 2

File: mne_denoise/viz/asr.py [hunk 10]
```diff
@@ -0,0 +1,820 @@
+"""Visualization helpers for Artifact Subspace Reconstruction (ASR).
```
Comment (BabaSanfour) [inline mne_denoise/viz/asr.py]: we changed the VIZ API from being method focused to being general, we removed all zapline and DSS specific functions. 

try to find ways to use functions we already have ; create new generic ones. same comment for all viz functions related to ASR

File: mne_denoise/utils.py [hunk 11]
```diff
@@ -50,7 +52,7 @@ def extract_data_from_mne(X: Any) -> tuple[np.ndarray, float | None, str, Any]:
     mne_type = "array"
     orig_inst = None
 
-    if _HAS_MNE and isinstance(X, BaseRaw | BaseEpochs | Evoked):
+    if _HAS_MNE and isinstance(X, (BaseRaw, BaseEpochs, Evoked)):  # noqa: UP038
```
Comment (BabaSanfour) [inline mne_denoise/utils.py]: remove the noqa if possible .. it skips checks

File: scripts/_fill_aasr_readme_numbers.py
Comment (BabaSanfour) [inline scripts/_fill_aasr_readme_numbers.py:1]: should be removed?

File: scripts/build_long_recording_substrate.py
Comment (BabaSanfour) [inline scripts/build_long_recording_substrate.py:1]: too many scripts / are they all needed? but not a priority we can sort them later.

File: tests/parity/matlab_reference/rasr_matlab_shims/grassmannfactory.m
Comment (BabaSanfour) [inline tests/parity/matlab_reference/rasr_matlab_shims/grassmannfactory.m:1]: should be removed / skipped once we are sure everything works

File: tests/parity/matlab_reference/ASR_REFERENCE_README.md
Comment (BabaSanfour) [inline tests/parity/matlab_reference/ASR_REFERENCE_README.md:1]: should be removed

File: tests/asr/test_adaptive.py
Comment (BabaSanfour) [inline tests/asr/test_adaptive.py:1]: tested should be reformatted to follow same style as we have for zapline and dss and icanclean ! + better commenting / doctsaings

**Review by BabaSanfour — CHANGES_REQUESTED** [review]

File: docs/asr/ASR_IMPLEMENTATION_PLAN.md
Comment (snesmaeili) [reply docs/asr/ASR_IMPLEMENTATION_PLAN.md:1]: Removed `docs/asr/ASR_IMPLEMENTATION_PLAN.md` -- it was an internal planning doc that should not ship with the package.

File: examples/tutorials/paper4_asr_kothe_chang.ipynb
Comment (snesmaeili) [reply examples/tutorials/paper4_asr_kothe_chang.ipynb:1]: Agreed -- removed `paper4_asr_kothe_chang.ipynb` from this branch; it belongs on the tutorials branch, not in the ASR PR. It stays in git history so it can be picked onto tutorials.

File: scripts/_fill_aasr_readme_numbers.py
Comment (snesmaeili) [reply scripts/_fill_aasr_readme_numbers.py:1]: Removed `scripts/_fill_aasr_readme_numbers.py` -- it was a one-off README number-filler, not part of the package.

File: tests/parity/matlab_reference/rasr_matlab_shims/grassmannfactory.m
Comment (snesmaeili) [reply tests/parity/matlab_reference/rasr_matlab_shims/grassmannfactory.m:1]: Removed the unused MATLAB shims (`grassmannfactory.m`, `manoptsolve.m`, `positive_definite_karcher_mean.m`); `generate_rasr_reference.m` now documents the Manopt prerequisite instead of vendoring shims.

File: tests/parity/matlab_reference/ASR_REFERENCE_README.md
Comment (snesmaeili) [reply tests/parity/matlab_reference/ASR_REFERENCE_README.md:1]: Removed `tests/parity/matlab_reference/ASR_REFERENCE_README.md` (stale internal notes).

File: docs/changes/devel/27.feature.rst
Comment (snesmaeili) [reply docs/changes/devel/27.feature.rst:1]: Consolidated into the single `docs/changes/devel/36.feature.rst` entry for this PR; deleted `27.feature.rst`.

File: docs/changes/devel/28.feature.rst
Comment (snesmaeili) [reply docs/changes/devel/28.feature.rst:1]: Consolidated into `docs/changes/devel/36.feature.rst`; deleted `28.feature.rst`.

File: docs/changes/devel/29.bugfix.rst
Comment (snesmaeili) [reply docs/changes/devel/29.bugfix.rst:1]: Consolidated into `docs/changes/devel/36.bugfix.rst`; deleted `29.bugfix.rst`.

File: docs/changes/devel/30.feature.rst
Comment (snesmaeili) [reply docs/changes/devel/30.feature.rst:1]: Consolidated into `docs/changes/devel/36.feature.rst`; deleted `30.feature.rst`.

File: docs/changes/devel/31.feature.rst
Comment (snesmaeili) [reply docs/changes/devel/31.feature.rst:1]: Consolidated into `docs/changes/devel/36.feature.rst`; deleted `31.feature.rst`.

File: mne_denoise/utils.py (hunk 11 above)
Comment (snesmaeili) [reply mne_denoise/utils.py]: Replaced the `# noqa` with a proper `isinstance(X, BaseRaw | BaseEpochs | Evoked)` union, so the lint suppression is gone.

File: examples/asr/plot_01_asr_basics.py (hunk 1 above)
Comment (snesmaeili) [reply examples/asr/plot_01_asr_basics.py]: Done -- `plot_01` now draws the before/after trace with `mne_denoise.viz.plot_signal_overlay`, then overlays the reference signal + burst shading on the returned axis.

File: examples/asr/plot_02_mne_raw_qc.py (hunk 2 above)
Comment (snesmaeili) [reply examples/asr/plot_02_mne_raw_qc.py]: Done -- `plot_02` now uses `plot_signal_overlay` for the before/after EEG trace; the repair/reject annotation spans are drawn on the returned axis.

File: mne_denoise/viz/asr.py (hunk 10 above)
Comment (snesmaeili) [reply mne_denoise/viz/asr.py]: Done -- `viz/asr.py` now keeps only the three ASR-specific diagnostics (repair timeline, calibration fraction, component reconstruction). Before/after, PSD and topomap views reuse the generic `plot_signal_overlay` / `plot_psd_comparison` / `plot_power_ratio_map`. Examples, docs and tests updated to match.

File: mne_denoise/asr/adaptive.py (hunk 3 above)
Comment (snesmaeili) [reply mne_denoise/asr/adaptive.py]: Expanded the `adaptive.py` module docstring: overview of the AASR variants (PSP/PSW/MW), references, and the public fit/partial_fit/transform API.

File: mne_denoise/asr/adaptive.py (hunk 4 above)
Comment (snesmaeili) [reply mne_denoise/asr/adaptive.py:57]: Made it consistent with the other modules: the MNE types are now guarded behind `try: from mne.io import BaseRaw ... except ImportError` with `Any` fallbacks, instead of a bare `import mne`.

File: mne_denoise/asr/adaptive.py (hunk 5 above)
Comment (snesmaeili) [reply mne_denoise/asr/adaptive.py]: Moved the helpers out: the Yule-Walker filter cluster now lives in `mne_denoise/asr/_aasr_filter.py` and the SPD/Riemannian primitives in `_spd.py`, so `adaptive.py` keeps just the estimator/pipeline logic.

File: mne_denoise/asr/adaptive.py (hunk 6 above)
Comment (snesmaeili) [reply mne_denoise/asr/adaptive.py:68]: No change needed for memory/speed: the adaptive path streams per window and the covariance aggregation is bounded by `max_mem_mb` (chunked geometric median), so peak memory does not grow with recording length; the per-window cost is O(C^3) in channel count, which dominates regardless.

File: mne_denoise/asr/core.py (hunk 7 above)
Comment (snesmaeili) [reply mne_denoise/asr/core.py]: Added explicit `meg`/`mag`/`grad` branches to `_resolve_picks` (with a clear error for unknown pick strings) and made the variance-floor check relative so Tesla-scale MEG data is not falsely rejected. Covered by new tests in `test_asr_coverage.py`.

File: mne_denoise/asr/core.py (hunk 8 above)
Comment (snesmaeili) [reply mne_denoise/asr/core.py]: Helpers: extracted to `_spd.py` (SPD primitives) and `_aasr_filter.py`, so `core.py` keeps just the pipeline. Performance: covariance aggregation is `max_mem_mb`-bounded (chunked geometric median) and the per-window eigendecomposition is O(C^3) in channels, so memory/time do not scale with recording length.

File: mne_denoise/asr/core.py (hunk 9 above)
Comment (snesmaeili) [reply mne_denoise/asr/core.py]: Split out -- the SPD/Riemannian block now lives in `mne_denoise/asr/_spd.py`; `core.py` imports just the primitives it uses.

File: tests/asr/test_adaptive.py
Comment (snesmaeili) [reply tests/asr/test_adaptive.py:1]: Reformatted `test_aasr_mw.py` to the zapline/dss/icanclean house style: simple imports with `assert_allclose`, `# ===` section banners, ASCII-only, one-line behavioral docstrings, and clearer inline comments.

File: scripts/build_long_recording_substrate.py
Comment (snesmaeili) [reply scripts/build_long_recording_substrate.py:1]: Acknowledged -- these are the paper-validation substrate/figure builder scripts behind the `reports/paper_validation` deliverables. Leaving them for now per your note; happy to prune/relocate the scripts tree in a follow-up.

File: docs/changes/devel/36.bugfix.rst
Comment (BabaSanfour) [inline docs/changes/devel/36.bugfix.rst:1]: to be removed

File: examples/asr/plot_01_asr_basics.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    pick=0,
+    scale_after=False,
+    before_label="Noisy",
+    after_label="ASR cleaned",
+    x_label="Time (s)",
+    y_label="Amplitude (a.u.)",
+    title="ASR burst repair",
+    show=False,
+)
+ax = fig.axes[0]
```
Comment (BabaSanfour) [inline examples/asr/plot_01_asr_basics.py]: This part can be added inside of the function plot_signal_overlay as optional

File: examples/asr/plot_02_mne_raw_qc.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    clean,
+    times,
+    scale_after=False,
+    before_label="Noisy EEG",
+    after_label="ASR cleaned EEG",
+    x_label="Time (s)",
+    y_label="Amplitude (a.u.)",
+    show=False,
+)
+ax = fig.axes[0]
```
Comment (BabaSanfour) [inline examples/asr/plot_02_mne_raw_qc.py]: same here?

File: mne_denoise/asr/core.py (hunk 9 above)
Comment (BabaSanfour) [reply mne_denoise/asr/core.py]: Its still around 3k lines; the helper functions at the end of the module could benefit from being moved somewhere else

File: tests/parity/matlab_reference/generate_rasr_reference.m
Comment (BabaSanfour) [inline tests/parity/matlab_reference/generate_rasr_reference.m:1]: delete

File: tests/asr/test_coverage.py
Comment (BabaSanfour) [inline tests/asr/test_coverage.py:1]: naming of test should follow scripts in main module .. this has to be renamed / moved

File: tests/asr/test_robustness.py
Comment (BabaSanfour) [inline tests/asr/test_robustness.py:1]: naming of test should follow scripts in main module .. this has to be renamed / moved

**Review by BabaSanfour — CHANGES_REQUESTED** [review]
The work is shaping nicely; one round of modifications before I take it from there

File: docs/changes/devel/36.bugfix.rst
Comment (snesmaeili) [reply docs/changes/devel/36.bugfix.rst:1]: Removed `36.bugfix.rst` and folded the `max_mem_mb` note into `36.feature.rst` (it's part of the new feature, not a fix to released behavior).

File: examples/asr/plot_01_asr_basics.py (hunk 12 above)
Comment (snesmaeili) [reply examples/asr/plot_01_asr_basics.py]: Moved the reference/highlight overlay into `plot_signal_overlay` as optional kwargs; the example now passes `reference=` and `highlight_mask=` directly instead of editing the axis.

File: examples/asr/plot_02_mne_raw_qc.py (hunk 13 above)
Comment (snesmaeili) [reply examples/asr/plot_02_mne_raw_qc.py]: Same here — `plot_02` now uses `reference=` and the new `highlight_spans=` kwarg of `plot_signal_overlay` instead of shading the axis by hand.

File: tests/parity/matlab_reference/generate_rasr_reference.m
Comment (snesmaeili) [reply tests/parity/matlab_reference/generate_rasr_reference.m:1]: Deleted. The rASR parity tests already `pytest.skip` when the (uncommitted) reference `.mat` is absent, so nothing depends on this generator.

File: tests/asr/test_coverage.py
Comment (snesmaeili) [reply tests/asr/test_coverage.py:1]: Renamed/moved: the tests were redistributed into the existing per-module suites (`test_estimator.py` / `test_adaptive.py` / `test_juggler.py`) by the module each exercises; `test_coverage.py` is deleted.

File: tests/asr/test_robustness.py
Comment (snesmaeili) [reply tests/asr/test_robustness.py:1]: Same — the parametric / edge-case / cross-variant tests now live in `test_estimator.py` / `test_adaptive.py` / `test_juggler.py`; `test_robustness.py` is deleted.

File: mne_denoise/asr/core.py (hunk 9 above)
Comment (snesmaeili) [reply mne_denoise/asr/core.py]: `core.py` is already a ~130-line re-export facade — the split landed in #40 (merged), and the helpers now live in `_estimator.py`, `_calibration.py`, `_covariance.py`, etc.

Comment (BabaSanfour) [reply mne_denoise/asr/core.py]: saw it and merged it mb

**Review by BabaSanfour — APPROVED** [review]

**Comment by BabaSanfour** [comment]
nice ! @snesmaeili I will take it from here
